File: Models/Password.py
# File = robjam1990/Psychosis/Models/Password.py
from math.matrix.Matrix import generate as generate_matrix, dot

# Password

MIN_PASSWORD_LENGTH = 8
MAX_PASSWORD_LENGTH = 16

def accept_password_input():
    password = []
    i = 0

    print(f'Enter your password (min {MIN_PASSWORD_LENGTH} characters): ')

    while True:
        ch = ord(getch())
        if ch == 13 or i >= MAX_PASSWORD_LENGTH - 1:
            break
        elif ch != 8:  # Check for backspace
            print('*', end='', flush=True)  # Print asterisks instead of actual characters
            password.append(chr(ch))
            i += 1
        elif i > 0:  # Handle backspace
            print('\b \b', end='', flush=True)  # Move cursor back, overwrite with space, move cursor back again
            i -= 1

    password.append('\0')  # Null terminate the password string

    if i < MIN_PASSWORD_LENGTH:
        print(f'\nPassword is too short. Please use at least {MIN_PASSWORD_LENGTH} characters.\n')
    else:
        print("\nPassword accepted. You can now proceed with the game.\n")

    return password

def encrypt(password):
    # Implementation to encrypt the password
    return password

alphabet_code_shift = ord('A')
english_alphabet_size = 26

def square_matrix_rotation(original_matrix):
    matrix = original_matrix[:]
    size = len(matrix)

    for i in range(size // 2):
        for j in range(i, size - i - 1):
            temp = matrix[i][j]
            matrix[i][j] = matrix[size - 1 - j][i]
            matrix[size - 1 - j][i] = matrix[size - 1 - i][size - 1 - j]
            matrix[size - 1 - i][size - 1 - j] = matrix[j][size - 1 - i]
            matrix[j][size - 1 - i] = temp

    return matrix

ZERO_OR_MORE_CHARS = '*'
ANY_CHAR = '.'

def regular_expression_matching(string, pattern):
    match_matrix = [[False] * (len(pattern) + 1) for _ in range(len(string) + 1)]
    match_matrix[0][0] = True

    for column_index in range(1, len(pattern) + 1):
        if pattern[column_index - 1] == '*':
            match_matrix[0][column_index] = match_matrix[0][column_index - 2]

    for row_index in range(1, len(string) + 1):
        for column_index in range(1, len(pattern) + 1):
            string_index = row_index - 1
            pattern_index = column_index - 1

            if pattern[pattern_index] == '.' or pattern[pattern_index] == string[string_index]:
                match_matrix[row_index][column_index] = match_matrix[row_index - 1][column_index - 1]
            elif pattern[pattern_index] == '*':
                match_matrix[row_index][column_index] = match_matrix[row_index][column_index - 2]
                if pattern[pattern_index - 1] == '.' or pattern[pattern_index - 1] == string[string_index]:
                    match_matrix[row_index][column_index] |= match_matrix[row_index - 1][column_index]
            else:
                match_matrix[row_index][column_index] = False

    return match_matrix[len(string)][len(pattern)]

def generate_key_matrix(key_string):
    matrix_size = int(len(key_string) ** 0.5)
    if matrix_size * matrix_size != len(key_string):
        raise ValueError('Invalid key string length. The square root of the key string must be an integer')

    key_string_index = 0
    return generate_matrix([matrix_size, matrix_size], lambda: (ord(key_string[key_string_index]) % alphabet_code_shift, key_string_index := key_string_index + 1)[0])

def generate_message_vector(message):
    return generate_matrix([len(message), 1], lambda cell_indices: ord(message[cell_indices[0]]) % alphabet_code_shift)

def hill_cipher_encrypt(message, key_string):
    only_letters_regexp = /^[a-zA-Z]+$/
    if not only_letters_regexp.match(message) or not only_letters_regexp.match(key_string):
        raise ValueError('The message and key string can only contain letters')

    key_matrix = generate_key_matrix(key_string)
    message_vector = generate_message_vector(message)

    if len(key_matrix) != len(message):
        raise ValueError('Invalid key string length. The key length must be a square of message length')

    cipher_vector = dot(key_matrix, message_vector)
    cipher_string = ''
    for item in cipher_vector:
        cipher_string += chr((item % english_alphabet_size) + alphabet_code_shift)

    return cipher_string

def hill_cipher_decrypt():
    raise NotImplementedError('This method is not implemented yet')

english_alphabet = list('abcdefghijklmnopqrstuvwxyz')

def get_cipher_map(alphabet, shift):
    return {current_char: alphabet[(char_index + shift) % len(alphabet)] for char_index, current_char in enumerate(alphabet)}

File: Models/Password.js
// File = robjam1990/Psychosis/Models/Password.js
import *as mtrx from '../../math/matrix/Matrix';

// Password

/**
 * Function to accept password input from the user.
 */
function acceptPasswordInput() {
    const password = [];
    let ch;
    let i = 0;

    console.log(`Enter your password(min ${MIN_PASSWORD_LENGTH} characters) : `);

    // Loop to read characters until Enter (ASCII 13) is pressed or maximum password length is reached
    while ((ch = getch()) != = 13 && i < MAX_PASSWORD_LENGTH - 1)
    {
        if (ch != = '\b')
        {                 // Check for backspace
            putchar('*'); // Print asterisks instead of actual characters
            password[i++] = ch;
        }
        else if (i > 0) {                    // Handle backspace
            printf("\b \b"); // Move cursor back, overwrite with space, move cursor back again
            i--;
        }
    }

    password[i] = '\0'; // Null terminate the password string

    // Check password length
    if (i < MIN_PASSWORD_LENGTH)
        console.log(`\nPassword is too short.Please use at least ${MIN_PASSWORD_LENGTH} characters.\n`);
    else
        console.log("\nPassword accepted. You can now proceed with the game.\n");

    return password;
}

/**
 * Function to encrypt the password.
 * @param {string} password - The password to encrypt.
 * @returns {string} - The encrypted password.
 */
function Encrypt(password) {
    // Implementation to encrypt the password
    return password;
}

export { acceptPasswordInput, Encrypt };

const alphabetCodeShift = 'A'.codePointAt(0);
const englishAlphabetSize = 26;

/**
 * Rotate a square matrix by 90 degrees clockwise.
 * @param {*[][]} originalMatrix - The original square matrix to rotate.
 * @return {*[][]} - The rotated square matrix.
 */
export default function squareMatrixRotation(originalMatrix) {
    const matrix = originalMatrix.slice();
    const size = matrix.length;

    for (let i = 0; i < Math.floor(size / 2); i++) {
        for (let j = i; j < size - i - 1; j++) {
            let temp = matrix[i][j];
            matrix[i][j] = matrix[size - 1 - j][i];
            matrix[size - 1 - j][i] = matrix[size - 1 - i][size - 1 - j];
            matrix[size - 1 - i][size - 1 - j] = matrix[j][size - 1 - i];
            matrix[j][size - 1 - i] = temp;
        }
    }

    return matrix;
}

const ZERO_OR_MORE_CHARS = '*';
const ANY_CHAR = '.';

/**
 * Determine if a string matches a pattern containing '*' and '.' wildcards.
 * @param {string} string - The input string.
 * @param {string} pattern - The pattern to match against.
 * @return {boolean} - Whether the string matches the pattern.
 */
/**
 * Determine if a string matches a pattern containing '*' and '.' wildcards.
 * @param {string} string - The input string.
 * @param {string} pattern - The pattern to match against.
 * @return {boolean} - Whether the string matches the pattern.
 */
export default function regularExpressionMatching(string, pattern) {
    const matchMatrix = Array(string.length + 1).fill(null).map(() = > {
        return Array(pattern.length + 1).fill(false);
    });

    matchMatrix[0][0] = true;

    for (let columnIndex = 1; columnIndex <= pattern.length; columnIndex++) {
        if (pattern[columnIndex - 1] == = '*')
        {
            matchMatrix[0][columnIndex] = matchMatrix[0][columnIndex - 2];
        }
    }

    for (let rowIndex = 1; rowIndex <= string.length; rowIndex++) {
        for (let columnIndex = 1; columnIndex <= pattern.length; columnIndex++) {
            const stringIndex = rowIndex - 1;
            const patternIndex = columnIndex - 1;

            if (pattern[patternIndex] == = '.' || pattern[patternIndex] == = string[stringIndex])
            {
                matchMatrix[rowIndex][columnIndex] = matchMatrix[rowIndex - 1][columnIndex - 1];
            }
            else if (pattern[patternIndex] == = '*')
            {
                matchMatrix[rowIndex][columnIndex] = matchMatrix[rowIndex][columnIndex - 2];
                if (pattern[patternIndex - 1] == = '.' || pattern[patternIndex - 1] == = string[stringIndex])
                {
                    matchMatrix[rowIndex][columnIndex] = matchMatrix[rowIndex][columnIndex] || matchMatrix[rowIndex - 1][columnIndex];
                }
            }
            else
            {
                matchMatrix[rowIndex][columnIndex] = false;
            }
        }
    }

    return matchMatrix[string.length][pattern.length];
}

/**
 * Generates key matrix from given keyString.
 * @param {string} keyString - a string to build a key matrix (must be of matrixSize^2 length).
 * @return {number[][]} keyMatrix
 */
const generateKeyMatrix = (keyString) = >
{
    const matrixSize = Math.sqrt(keyString.length);
    if(!Number.isInteger(matrixSize))
    {
    throw new Error('Invalid key string length. The square root of the key string must be an integer');
}
let keyStringIndex = 0;
return mtrx.generate(
    [matrixSize, matrixSize],
    () = > {
        const charCodeShifted = (keyString.codePointAt(keyStringIndex)) % alphabetCodeShift;
        keyStringIndex += 1;
        return charCodeShifted;
    },);
};

/**
 * Generates a message vector from a given message.
 * @param {string} message - the message to encrypt.
 * @return {number[][]} messageVector
 */
const generateMessageVector = (message) = >
{
    return mtrx.generate(
        [message.length, 1],
        (cellIndices) = > {
            const rowIndex = cellIndices[0];
            return message.codePointAt(rowIndex) % alphabetCodeShift;
        },);
};

/**
 * Encrypts the given message using Hill Cipher.
 * @param {string} message - The plaintext message to encrypt.
 * @param {string} keyString - The key to use for encryption.
 * @return {string} - The encrypted cipher string.
 */
export function hillCipherEncrypt(message, keyString) {
    const onlyLettersRegExp = / ^[a - zA - Z] + $ /;
    if (!onlyLettersRegExp.test(message) || !onlyLettersRegExp.test(keyString)) {
        throw new Error('The message and key string can only contain letters');
    }

    const keyMatrix = generateKeyMatrix(keyString);
    const messageVector = generateMessageVector(message);

    if (keyMatrix.length != = message.length)
    {
        throw new Error('Invalid key string length. The key length must be a square of message length');
    }

    const cipherVector = mtrx.dot(keyMatrix, messageVector);
    let cipherString = '';
    for (let row = 0; row < cipherVector.length; row += 1) {
        const item = cipherVector[row];
        cipherString += String.fromCharCode((item % englishAlphabetSize) + alphabetCodeShift);
    }

    return cipherString;
}

/**
 * Decrypts the given message using Hill Cipher.
 * @return {string} - The decrypted plaintext message.
 */
export const hillCipherDecrypt = () = >
{
    throw new Error('This method is not implemented yet');
};

const englishAlphabet = 'abcdefghijklmnopqrstuvwxyz'.split('');

/**
 * Generate a cipher map based on the alphabet and a given shift.
 * @param {string[]} alphabet - The array representing the alphabet.
 * @param {number} shift - The shift for encryption or decryption.
 * @return {Object} - The cipher map.
 */
const getCipherMap = (alphabet, shift) = >
{
    return alphabet.reduce((charsMap, currentChar, charIndex) = > {
        let encryptedCharIndex = (charIndex + shift) % alphabet.length;
        if(encryptedCharIndex < 0) {
            encryptedCharIndex += alphabet.length;
        }
        charsMap[currentChar] = alphabet[encryptedCharIndex];
        return charsMap;
    },
        {});
};
